README.md
# 💡 delux 💡

[![CircleCI](https://circleci.com/gh/smartrent/delux.svg?style=svg)](https://circleci.com/gh/smartrent/delux)
[![Hex version](https://img.shields.io/hexpm/v/delux.svg "Hex version")](https://hex.pm/packages/delux)

_de lux_ | _Latin (roughly) for "of the light"_

<!-- MODULEDOC -->
Use LEDs for your user interface

`delux` simplifies creating and running LED blink patterns for use as part of a
user interface for an embedded hardware. This library provides:

* Low overhead LED control via Linux's Sysclass interface - blink sequences
  get compiled down so that they can be run inside the Linux kernel.
* Built-in LED effects for common use cases like blinking, cycling LED colors,
  etc.
* Prioritization of LED effects
* Support for many physical LED configurations
* Nice textual descriptions of what the LED is doing to support remote debug

This library is primarily intended for devices with 1 to 10 LEDs. It's
currently not for Neopixels and other "smart" LEDs that are more typically
found in larger numbers, but could be used for a similar purpose.

Before diving in, some terminology is needed:

* LED - one light emitting element. This doesn't have to be an LED, but the
  Linux kernel must think that it is and show a directory for it under
  `/sys/class/leds/`.
* Indicator - a group of 1, 2, or 3 LEDs that a user would perceive as one.
  This could be a lone green LED, or a red, green and blue LED in one package,
  or any combination.
* Program - a one-time or repeating set of instructions for controlling an
  indicator.
* Pattern - a low-level sequence of brightness and duration tuples for
  controlling one LED
* Slot - a holder for a program. Slots are ordered so a program put in a higher
  priority slot will take precedence over one in a lower priority slot. For
  example, if there's a UI feedback slot and a network status slot, programs
  running in the UI feedback slot could take precedence.

To give a flavor of how `delux` works, here's an example that configures
`delux` with one green LED and then blinks it at 2 Hz:

```elixir
iex> Delux.start_link(indicators: %{default: %{green: "led0"}})
iex> Delux.render(Delux.Effects.blink(:green, 2))
iex> Delux.info()
green at 2 Hz
```

This starts Delux with one indicator, `:default`, that has a green LED known to
Linux as `"led0"`. The `Delux.Effects.blink/2` function creates a 2 Hz blinking
program for Delux to render. With nothing else specified, `Delux.render/1` runs
the program on the default indicator in the default slot.

<!-- MODULEDOC -->

## Built-in indicator programs

`delux` comes with helper functions for creating LED programs. It's also
possible to create your own. Here are the built-in ones:

| Function                    | Description  |
| --------------------------- | ------------ |
| `Delux.Effects.off/0`       | Turn an indicator off |
| `Delux.Effects.on/2`        | Set the indicator to the specified color (within reason, that is. You can't make a green LED red, for example.) |
| `Delux.Effects.blink/3`     | Blink a color on and off at a fixed rate |
| `Delux.Effects.number_blink/3` | Blink the indicator the specified number of times |
| `Delux.Effects.blip/3`      | Show two colors quickly in succession. E.g., use to show feedback from pressing a button. |
| `Delux.Effects.cycle/3`     | Cycle through a set of colors at a fixed rate |
| `Delux.Effects.waveform/2`  | Synthesize an LED pattern from a function |
| `Delux.Morse.encode/3`      | Send a string in Morse code |

## Before you use `delux`

It's worthwhile thinking about how you want your device's LEDs to behave before
writing any code.

The first step is to figure out what LEDs are available and how to group them
into indicators. To see what's available, list the directories in
`/sys/class/leds`. If something is missing, you'll likely need to adjust the
device tree configuration and make it appear. That's discussed somewhat below,
but this is device-specific, so you'll need to look elsewhere for precise
guides. After you've found the LEDs, group them and give them indicator names.
If you just have one indicator, call it `:default`. That will make `delux`'s
API more convenient.

Step 2 is to decide what slots make sense for your application. Only one
program per indicator can be run at a time. Every time that you set a program
on an indicator, it replaces any running programs for that indicator in that
slot. The default list of slots probably suffice to start:

* `:status` - The lowest priority slot. This is for general device status like
  whether networking is working and if the device is initializing and connected
  to the back end.
* `:notification` - This is a medium priority slot. Use it to show transient
  things like an alert that requires operator attention to clear.
* `:user_feedback` - This is the highest priority slot and for showing feedback
  to a user. For example, it could blink the LED when user pushes a button so
  they know that the device is doing something.

Clearing the program in a slot makes `delux` render the program on the next
lower priority slot or if there's no program, then the indicator is turned off.

> #### Tip {: .tip}
>
> If you find yourself creating state machines in your code to control the
> indicators, try adding one or more slots as an alternative.

Once you feel good about the indicators and slots, it's time to configure
`delux`.

## Configuration

After adding `delux` to your mix dependencies, you can either add `Delux` to a
supervision tree of your choice or add a `:delux` configuration to your
application environment (`config.exs`). The tradeoff is that having `Delux`
start itself by specifying an application config can be really convenient, but
adding `Delux` to your supervision tree allows runtime configuration and more
control over failure recovery.

Starting with the supervision tree approach, here's an example childspec:

```elixir
  {Delux,
   name: MyIndicators,
   indicators: %{
     default: %{red: "led0:red", green: "led0:green", blue: "led0:blue"},
     indicator2: %{red: "led1"}
   }}
```

The above configuration shows two indicators. The first is called `:default` and
is an RGB indicator. The second is a lone red LED that's used as `:indicator2`.
As mentioned before, if you only have one indicator, call it `:default`.

Other options include setting the list of slots and giving the `Delux` GenServer
a name. If you don't give the `Delux` GenServer a name, it will register itself
as a singleton and you won't have to pass the server name or pid to any of the
API calls.

If you'd like Delux to start itself automatically, add the configuration to your
`config.exs` or a file that it includes. Here's an example:

```elixir
config :delux,
   indicators: %{
     default: %{red: "led0:red", green: "led0:green", blue: "led0:blue"},
     indicator2: %{red: "led1"}
   }}
```

## Use

For sake of example, let's start the `Delux` GenServer the manual way by calling
`start_link/1` directly. Modify the LED names to whatever you have.

```elixir
iex> Delux.start_link(indicators: %{
     default: %{red: "led0:red", green: "led0:green", blue: "led0:blue"},
     indicator2: %{red: "led1"}
   })
```

After you have a `Delux` GenServer and running, call `Delux.render/1` to turn on
the default indicator in the default slot:

```elixir
iex> Delux.render(Delux.Effects.on(:white))
:ok
```

Even if you don't have an RGB indicator, passing `:white` should turn on all of
the LEDs that you do have.

If you have two indicators and want them to blink back and forth, you can try
this:

```elixir
iex> Delux.render(%{
    default: Delux.Effects.cycle([:black, :white], 1),
    indicator2: Delux.Effects.cycle([:white, :black], 1)
  })
:ok
```

Finally, pass a second parameter to `Delux.render/2` to assign the programs to
another slot.

## Creating your own programs

Indicator programs are a collection of LED patterns and metadata as defined by
the `Delux.Program` struct. The helper functions in `Delux.Effects` and
`Delux.Morse` create these and can be used as examples.

The core of the program is an `Delux.Pattern`. There are three patterns per
program for each of the color channels.

Each pattern is a list of `{value, duration}` tuples where `value` is a number
from 0 to 1 and duration is an integer number of milliseconds. While `delux`
internally holds a number from 0 (off) to 1 (full on) for the LED's value at
that point in time, this gets scaled to an integer when sent to Linux. This
integer depends on the maximum brightness value for an LED. This is often just
1 since Linux can only turn the LED off and on. Sometimes Linux can set the LED
to multiple levels and in those cases, the maximum value will be greater than 1
(often 255). You don't need to worry about scaling other than to be aware that
fractional values will be rounded when sent to Linux and this will mess up
colors.

Linux interpolates the LED value over the duration between pairs of tuples. A
common idiom is to have 0 duration tuples to turn off interpolation. For
example, `[{1, 100}, {1, 0}, {0, 100}, {0, 0}]` turns the LED on for 100 ms and
then off for 100 ms without any interpolation.

See Linux's LED pattern trigger documentation at
[leds-trigger-pattern.txt](https://elixir.bootlin.com/linux/v5.19/source/Documentation/devicetree/bindings/leds/leds-trigger-pattern.txt)
for more info.

## Timing precision

The Linux kernel's `HZ` configuration sets the granularity at which Delux can
render patterns. For example, if `HZ=100`, the timing resolution is 10 ms. The
kernel also adds a start delay for any pattern. Delux can compensate for both
effects to more accurately render programs. This is ultimately limited to the
timer resolution so if the system only supports 10 ms resolution, that's the
best that Delux can do.

Delux can't automatically detect the `HZ` setting so it must be told via the
`:backend` options:

```elixir
Delux.start_link(backend: [hz: 100], indicators: %{default: %{green: "led0"}})
```

If you need better resolution, you'l need to update the Linux kernel
configuration with `CONFIG_HZ=1000` and rebuild. Be sure to update the `:hz`
value passed to `Delux` so that the compensation is calculated correctly.

To verify timing, connect a logic analyzer to an LED controlled by Delux and
run this:

```elixir
Delux.render(Delux.Effects.timing_test(:on))
```

See `Delux.Effects.timing_test/2` for details on what you should see.

## Linux kernel configuration

`delux` works in all official Nerves systems. If not using Nerves, you will
need to have `CONFIG_LEDS_TRIGGER_PATTERN=y` enabled your Linux kernel
configuration.

The second step to using Linux's LED subsystem is to configure LEDs in the
device tree. You can't just use an arbitrary GPIO to turn on the LED like you
can with `Circuits.GPIO`. Linux needs to know about the GPIO. It's also
possible to hook up PWMs and LED drivers. See the [LED
drivers](https://elixir.bootlin.com/linux/latest/source/drivers/leds) for
options.

> #### Tip {: .tip}
>
> If you're using a BeagleBone, check out
> [Udo Schneider's blog post on device tree overlays](https://pubray.com/udo-schneider/custom-device-tree-overlays-for-beagle-bone-black-running-nerves)

The following is an example device tree configuration for telling Linux about
GPIO-connected LEDs. It is platform-specific so you can't just copy/paste it.
It sets up two RGB LEDs and makes one blink early in the boot process.

```dts
/ {
        leds {
                pinctrl-names = "default";
                pinctrl-0 = <&pinctrl_leds>;
                compatible = "gpio-leds";

                led1_red {
                        label = "led1:red";
                        gpios = <&pio 0 7 GPIO_ACTIVE_LOW>;
                };
                led1_green {
                        label = "led1:green";
                        gpios = <&pio 4 15 GPIO_ACTIVE_LOW>;
                };
                led1_blue {
                        label = "led1:blue";
                        gpios = <&pio 4 4 GPIO_ACTIVE_LOW>;

                        /* Blink LED at 2 Hz (250 ms on, off) */
                        linux,default-trigger = "timer";
                        default-state = "on";
                        led-pattern = <250 250>;
                };
                led2_red {
                        label = "led2:red";
                        gpios = <&pio 4 5 GPIO_ACTIVE_LOW>;
                };
                led2_green {
                        label = "led2:green";
                        gpios = <&pio 4 14 GPIO_ACTIVE_LOW>;
                };
                led2_blue {
                        label = "led2:blue";
                        gpios = <&pio 4 7 GPIO_ACTIVE_LOW>;
                };
        };
};
```

## License

```text
Copyright (C) 2022 SmartRent

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
```


CHANGELOG.md
# Changelog

## v0.4.1 - 2022-10-17

* Changes
  * Added support for starting Delux as an OTP application. This is triggered by
    supplying an application configuration (see README.md). Please see
    documentation for tradeoffs.
  * Added support for specifying an initial configuration. The indicators no
    longer are forced to start in the off state.
  * Support looping Morse code sequences

## v0.4.0 - 2022-10-14

This is a big update that improves program playback timing to compensate for
the Linux kernel's HZ configuration and implementation of LED pattern triggers.
This affects all programs, but you probably won't notice it unless you're
creating Morse code patterns or measuring LED output with a logic analyzer.

* Backwards incompatible changes
  * "Priority" is now called "slot". If you have custom priorities defined in
    your configurations, you will need to change them to `slot`. This change
    was made to make the use of this feature easier to explain. I.e. programs
    are put in slots, one program per slot per indicator, slots have an order.

* Changes
  * Added `:backend` parameters and specifically the `:hz` parameter to support
    better playback timing. If unset, HZ=1000 is assumed. This results in
    almost the same timings as earlier versions. See the README for more
    information.
  * Added `Delux.Effects.number_blink/3` for blinking out a number. This is for
    things like error codes - like error 1, 2, 3, etc.
  * Added `Delux.Effects.timing_test/2` to debug timing precision issues with a
    logic analyzer.
  * Removed unnecessary writes to the filesystem when using multiple
    indicators. This fixes an issue where a change on one indicator restarts a
    program on a second indicator.
  * Reorganized the backend code to support future alternative backends. This is
    currently unplanned work, but it's now easier to think about adding a way to
    run Delux using Circuits.GPIO or a simulated LED.

## v0.3.1 - 2022-09-26

* Changes
  * Support dimming arbitrary patterns to off. The dimming feature still doesn't
    dim all patterns, but dimming to off works for everything now.

## v0.3.0 - 2022-09-07

* Changes
  * Favor singleton use of Delux with default arguments to `render/3` and
    `info/2`. Now that Delux is being used more frequently as a singleton, this
    makes the API more convenient. This is backwards incompatible if you had
    been taking advantage of default arguments, but you should get compile-time
    and dialyzer errors. If you do name your Delux GenServers, you'll have to
    specify all arguments to `render/3` and `info/2`.
  * Improve argument error messages by reporting what the valid options are in
    addition to what you passed.

## v0.2.0 - 2022-08-26

* Changes
  * Register the Delux GenServer with a default name so that pids or names don't
    need to be passed to all APIs. Since Delux is almost always used as a
    singleton, this simplifies the API.

## v0.1.3 - 2022-08-11

* Changes
  * Fix typespecs to remove warnings found by Dialyzer
  * Update `Delux.Effects.waveform/3` to support color atom names and check RGB
    tuples for range. The latter fixes errors that get detected later on and are
    more confusing to figure out.

## v0.1.2 - 2022-08-11

* Changes
  * Improve `Delux.Effects.waveform/3` so that it's easier to use and produces
    shorter patterns. It now has examples in the docs and unit tests.

## v0.1.1 - 2022-08-10

* Changes
  * Adjust timed pattern playback to minimize trimming LED programs.
    Unfortunately, LED programs still get cut off at the end with this release.
    If you're using Morse code, you'll see this. We plan on fixing this
    completely in a future release.
  * Support empty LED setups to simplify configuration of multi-target projects
    and unit tests.

## v0.1.0 - 2022-08-08

Initial release


lib/delux/effects.ex
defmodule Delux.Effects do
  @moduledoc """
  Functions for creating a variety of LED patterns
  """

  alias Delux.Pattern
  alias Delux.Program
  alias Delux.RGB

  @typedoc """
  Options for all effects (none yet)
  """
  @type common_options() :: []

  @doc """
  All LEDs off
  """
  @spec off() :: Program.t()
  def off() do
    %Program{
      red: led_off(),
      green: led_off(),
      blue: led_off(),
      description: "off",
      mode: :simple_loop
    }
  end

  @doc """
  Set an indicator to the specified color
  """
  @spec on(RGB.color(), common_options()) :: Program.t()
  def on(c, _options \\ []) do
    {r, g, b} = RGB.new(c)

    %Program{
      red: led_on(r),
      green: led_on(g),
      blue: led_on(b),
      description: RGB.to_ansidata(c, "solid "),
      mode: :simple_loop
    }
  end

  @doc """
  Blink an indicator

  This returns a pattern that blinks the specified color at a 50% duty cycle.
  The pattern starts on and then goes off.
  """
  @spec blink(RGB.color(), number(), common_options()) :: Program.t()
  def blink(c, frequency, _options \\ []) do
    {r, g, b} = RGB.new(c)

    %Program{
      red: led_blink(r, frequency),
      green: led_blink(g, frequency),
      blue: led_blink(b, frequency),
      description: [RGB.to_ansidata(c, ""), " at #{frequency} Hz"],
      mode: :simple_loop
    }
  end

  @doc """
  Create a transient two color sequence

  The first color is shown for 20 ms. 10 ms in, the second color is shown for
  20 ms. The Effects is a quick flash of light that can be used to show
  feedback to a button. Total duration of the Effects is 40 ms.
  """
  @spec blip(RGB.color(), RGB.color(), common_options()) :: Program.t()
  def blip(c1, c2, _options \\ []) do
    {r1, g1, b1} = RGB.new(c1)
    {r2, g2, b2} = RGB.new(c2)

    %Program{
      red: led_blip(r1, r2),
      green: led_blip(g1, g2),
      blue: led_blip(b1, b2),
      description: [RGB.to_ansidata(c1), :reset, "->", RGB.to_ansidata(c2), " blip"],
      mode: :one_shot
    }
  end

  @doc """
  Create a program to verify timing

  If you're unsure about the playback timing on your device, hook up a logic
  analyzer to an LED and capture the waveform. Here's what you should see:

  1. 100 ms on
  2. 10 ms off
  3. 1 ms on
  4. 10 ms off
  5. 2 ms on
  6. 10 ms off
  7. 3 ms on
  8. 10 ms off
  9. 5 ms on
  10. 10 ms off
  11. 8 ms on
  12. 10 ms off
  13. 13 ms on
  14. 10 ms off
  15. 100 ms on
  16. off

  Look at the following in the capture:

  1. Do the 1, 2, 3, 5 ms captures match what you'd expect based on your
     kernel's HZ setting. For example, if HZ=100, they should all be 10 ms long.
     If not, check the :hz setting for Delux.
  2. Does the length of the final 100 ms on time match the first. If not,
     Delux might be calculating the duration wrong and cutting off the program
     prematurely. This is likely due to an incorrect :hz setting.

  If something still isn't right, please submit an issue with the captured
  waveform and any other hints you may have to reproduce.
  """
  @spec timing_test(RGB.color(), common_options()) :: Program.t()
  def timing_test(c, _options \\ []) do
    {r, g, b} = RGB.new(c)

    %Program{
      red: led_timing(r),
      green: led_timing(g),
      blue: led_timing(b),
      description: ["Timing test pattern in ", RGB.to_ansidata(c)],
      mode: :one_shot
    }
  end

  defp led_timing(0), do: led_off()

  defp led_timing(b),
    do: [
      {b, 100},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 1},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 2},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 3},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 5},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 8},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 13},
      {b, 0},
      {0, 10},
      {0, 0},
      {b, 100},
      {b, 0}
    ]

  @doc """
  Cycle between colors

  Colors are shown with equal duration determined from the specified frequency.
  """
  @spec cycle([RGB.color()], number(), common_options()) :: Program.t()
  def cycle(colors, frequency, _options \\ [])
      when is_list(colors) and frequency > 0 and frequency < 20 do
    {reds, greens, blues} = colors |> Enum.map(&RGB.new/1) |> unzip3()

    duration = round(1000.0 / Enum.count(colors) / frequency)

    %Program{
      red: led_cycle(reds, duration),
      green: led_cycle(greens, duration),
      blue: led_cycle(blues, duration),
      description: [
        Enum.map(colors, &RGB.to_ansidata/1) |> Enum.intersperse([:reset, "-"]),
        :reset,
        " cycle at #{frequency} Hz"
      ],
      mode: :simple_loop
    }
  end

  @doc """
  Create a program from an arbitrary function

  Pass in a function that takes times in milliseconds and returns colors. The
  returned pattern piecewise linearly interpolates the waveform.

  Here's an example of a 0.5 Hz blue sine wave:

  ```elixir
  Effects.waveform(fn t -> {0, 0, 0.5 + 0.5 *:math.cos(:math.pi() * t /  1000)} end, 2000)
  ```

  When trying this, keep in mind that if the LEDs in the indicator don't support
  varying levels of brightness, it won't look like a sine wave.

  Options

  * `:time_step` - the number of milliseconds between each sample. Defaults to 100 ms.
  """
  @spec waveform((Pattern.milliseconds() -> RGB.color()), Pattern.milliseconds(), keyword()) ::
          Program.t()
  def waveform(fun, period, options \\ []) do
    time_step = options[:time_step] || 100
    colors = for t <- 0..period//time_step, do: RGB.new(fun.(t))
    {reds, greens, blues} = unzip3(colors)

    %Program{
      red: led_waveform(reds, time_step, period),
      green: led_waveform(greens, time_step, period),
      blue: led_waveform(blues, time_step, period),
      description: "waveform",
      mode: :simple_loop
    }
  end

  @typedoc false
  @type number_blink_options() :: [
          blink_on_duration: pos_integer(),
          blink_off_duration: pos_integer(),
          inter_number_delay: pos_integer()
        ]

  @doc """
  Blink out a number

  This returns a pattern that blinks out a number. It's good for
  communicating small numbers to viewers. It repeats.

  Options:

  * `:inter_number_delay` - the amount of milliseconds to wait in between
    blinking out the count (defaults to 2000 ms)
  * `:blink_on_duration` - how long to keep the LED on when blinking (defaults to 250 ms)
  * `:blink_off_duration` - how long to keep the LED off when blinking (defaults to 250 ms)
  """
  @spec number_blink(RGB.color(), 1..20, number_blink_options()) :: Program.t()
  def number_blink(c, count, options \\ []) when count >= 1 and count <= 20 do
    {r, g, b} = RGB.new(c)

    on = Keyword.get(options, :blink_on_duration, 250)
    off = Keyword.get(options, :blink_off_duration, 250)
    inter = Keyword.get(options, :inter_number_delay, 2000)

    %Program{
      red: led_number_blink(r, count, on, off, inter),
      green: led_number_blink(g, count, on, off, inter),
      blue: led_number_blink(b, count, on, off, inter),
      description: ["Blink ", RGB.to_ansidata(c, ""), " #{count} times"],
      mode: :simple_loop
    }
  end

  defp led_number_blink(0, _count, _on, _off, _inter), do: led_off()

  defp led_number_blink(b, count, on_time, off_time, inter) do
    [List.duplicate([{b, on_time}, {b, 0}, {0, off_time}, {0, 0}], count), {0, inter}, {0, 0}]
    |> List.flatten()
  end

  defp unzip3(tuples, acc \\ {[], [], []})
  defp unzip3([], {a1, a2, a3}), do: {Enum.reverse(a1), Enum.reverse(a2), Enum.reverse(a3)}
  defp unzip3([{x, y, z} | rest], {a1, a2, a3}), do: unzip3(rest, {[x | a1], [y | a2], [z | a3]})

  defp led_off(), do: [{0, Pattern.forever_ms()}, {0, 0}]
  defp led_on(b), do: [{b, Pattern.forever_ms()}, {b, 0}]
  defp led_blink(0, _frequency), do: led_off()

  defp led_blink(b, frequency) when frequency > 0 and frequency < 20 do
    period = round(1000 / frequency)
    on_time = div(period, 2)
    off_time = period - on_time
    [{b, on_time}, {b, 0}, {0, off_time}, {0, 0}]
  end

  defp led_blink(b, frequency) when frequency >= 20, do: led_on(b)

  defp led_blip(0, 0), do: led_off()
  defp led_blip(0, b2), do: [{0, 20}, {0, 0}, {b2, 20}, {b2, 0}]
  defp led_blip(b1, 0), do: [{0, 10}, {0, 0}, {b1, 20}, {b1, 0}]
  defp led_blip(b1, b2), do: [{0, 10}, {0, 0}, {b1, 30}, {b2, 0}]

  defp led_cycle(values, duration) do
    Enum.flat_map(values, fn b -> [{b, duration}, {b, 0}] end)
  end

  defp led_waveform(values, time_step, total_time) do
    {result, 0} =
      Enum.map_reduce(values, total_time, fn v, time_left ->
        time = min(time_step, time_left)
        {{v, time}, time_left - time}
      end)

    Pattern.simplify(result)
  end
end
